src/ctap/credential_id.rs

```diff
@@ -0,0 +1,227 @@
+        // we've already checked that the HMAC matched.
+        return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+    }
+    let pad_length = *data.last().unwrap();
+    if pad_length == 0 || pad_length > MAX_PADDING_LENGTH {
+        return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+    }
+    if !data
+        .drain((data.len() - pad_length as usize)..)
+        .all(|x| x == pad_length)
+    {
+        return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+    }
+    Ok(())
+}
+
+impl ParsedCredentialId {
+    /// Encrypts the given private key, relying party ID hash, and cred protect policy into a credential ID.
+    ///
+    /// Other information, such as a user name, are not stored. Since encrypted credential IDs are
+    /// stored server-side, this information is already available (unencrypted).
+    pub(super) fn encrypt(
+        env: &mut impl Env,
+        private_key: &PrivateKey,
+        rp_id_hash: &[u8; 32],
+        cred_protect_policy: Option<CredentialProtectionPolicy>,
+    ) -> Result<Vec<u8>, Ctap2StatusCode> {
+        let mut payload = Vec::new();
+        let cbor = cbor_map_options! {
+          CborKeyHandleField::PrivateKey => private_key,
+          CborKeyHandleField::RpIdHash=> rp_id_hash,
+          CborKeyHandleField::CredProtectPolicy => cred_protect_policy,
+        };
+        cbor_write(cbor, &mut payload)?;
+        add_padding(&mut payload)?;
+
+        let aes_enc_key =
+            crypto::aes256::EncryptionKey::new(&env.key_store().key_handle_encryption()?);
+        let encrypted_payload = aes256_cbc_encrypt(env.rng(), &aes_enc_key, &payload, true)?;
+        let mut credential_id = encrypted_payload;
+        credential_id.insert(0, CBOR_CREDENTIAL_ID_VERSION);
+
+        let id_hmac = hmac_256::<Sha256>(
+            &env.key_store().key_handle_authentication()?,
+            &credential_id[..],
+        );
+        credential_id.extend(&id_hmac);
+        Ok(credential_id)
+    }
+
+    /// Returns None if
+    /// - the format does not match any known versions, or
+    /// - the HMAC test fails.
+    ///
+    /// For v0 (legacy U2F) the credential ID consists of:
+    /// - 16 bytes: initialization vector for AES-256,
+    /// - 32 bytes: encrypted ECDSA private key for the credential,
+    /// - 32 bytes: encrypted relying party ID hashed with SHA256,
+    /// - 32 bytes: HMAC-SHA256 over everything else.
+    ///
+    /// For v1 (CBOR key handle) the credential ID consists of:
+    /// -   1 byte : version number
+    /// -  16 bytes: initialization vector for AES-256,
+    /// - 192 bytes: encrypted CBOR key handle
+    /// -  32 bytes: HMAC-SHA256 over everything else.
+    pub(super) fn decrypt(
+        env: &mut impl Env,
+        credential_id: &[u8],
+    ) -> Result<Option<Self>, Ctap2StatusCode> {
+        if credential_id.len() < MIN_CREDENTIAL_ID_SIZE {
```

Comment:
This function doesn't perform a length check for `decrypt_cbor_key_handle`. That shouldn't be a problem in practise, since we never HMAC anything of a different length, and worst case we fail decryption if it's not a multiple of block size.

I was wondering if we rather want to be more defensive, or keep this function open to accept unexpected credential IDs.

Reply:
I agree that I was quite confused by the difference between `Ok(None)` and `Err(_)`. It would be nice if we had a clear story. In particular, `Ok(None)` should not be used for errors. So it's all a question of what do we consider as errors. And I guess if something produced by us looks wrong it's an error.

Reply:
I forgot this open conversation, and went back to double check. My new comment is the only piece of code that I'm not sure is correct, WDYT?

Reply:
Let me look into it deeper and open a discussion if needed, let's merge this first.